### PR TITLE
[SPARK-36371][SQL] Support raw string literal

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR proposes to support raw string literal which escape no character using `\`.
The raw string literal is the form of `r"..."` or `r'...'` like the syntax BigQuery and Python supports.

Actually, there is no standard way to represent raw string literals.

In PostgreSQL, any special character isn't escaped by \ unless a string literal starts with E prefix.
https://www.postgresql.org/docs/13/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS

In MySQL, special characters in a string literal are not escaped if NO_BACKSLASH_ESCAPES is enabled.
https://dev.mysql.com/doc/refman/8.0/en/string-literals.html

In MsSQLServer, any special character isn't escaped by \ but STRING_ESCAPE function can escape such characters.
https://docs.microsoft.com/en-us/sql/t-sql/functions/string-escape-transact-sql?view=sql-server-ver15

But BigQuery supports `r"..."` and `r'...'` forms so this PR proposes this syntax for the purpose.
https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#literals

### Why are the changes needed?

In the current master, sometimes it's too confusable to represent JSON and regex in a string literal if they contain backslash.
For example, in JSON, `\` needs to be escaped like as follows.
```
{"a": "\\"}
```
But, if the JSON above is represented in a string literal, further two `\` are needed because string literal also requires `\` to be escaped.
```
SELECT from_json('{"a": "\\\\"}', 'a string')
{"a":"\"}
```
With the raw string literal, we can represent such JSON like as follows.
```
SELECT from_json(r'{"a": "\\"}', 'a string')
{"a":"\"}
```

### Does this PR introduce _any_ user-facing change?

No. This PR just extends the existing syntax.

### How was this patch tested?

Added new test.
I also confirmed that the modified document is successfully built with `SKIP_API=1 bundle exec jekyll build`.
raw_string_literal

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
retest this please

File: docs/sql-ref-literals.md [hunk 1]
```diff
@@ -47,6 +47,10 @@ A string literal is used to specify a character string value.
     To represent unicode characters, use 16-bit or 32-bit unicode escape of the form `\uxxxx` or `\Uxxxxxxxx`,
     where xxxx and xxxxxxxx are 16-bit and 32-bit code points in hexadecimal respectively (e.g., `\u3042` for `あ` and `\U0001F44D` for `👍`).
 
+* **R**
```
Comment (HyukjinKwon) [inline docs/sql-ref-literals.md:50]: Hm .. I wonder if other DBMSes have their own syntax for this that we can refer to instead of the Python's ...

Comment (sarutak) [reply docs/sql-ref-literals.md:50]: Actually, I investigated raw string syntax support in other RDBMS beforehand, and I found there is no standard way to represent raw string literals.

In PostgreSQL, any special character isn't escaped by `\` unless a string literal starts with `E` prefix.
https://www.postgresql.org/docs/13/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS

In MySQL, special characters in a string literal are not escaped if `NO_BACKSLASH_ESCAPES` is enabled.
https://dev.mysql.com/doc/refman/8.0/en/string-literals.html

In MsSQLServer, any special character isn't escaped by `\` but `STRING_ESCAPE` function can escape such characters.
https://docs.microsoft.com/en-us/sql/t-sql/functions/string-escape-transact-sql?view=sql-server-ver15

It's not a RDBMS but BigQuery supports raw string literals like Python.
https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#literals

So, I choose this syntax.

Comment (sarutak) [reply docs/sql-ref-literals.md:50]: I've updated the description to explain why `r'...'` and `r"..."` syntax is chosen.

**Comment by sarutak** [comment]
cc: @cloud-fan

File: docs/sql-ref-literals.md (hunk 1 above)
Comment (cloud-fan) [inline docs/sql-ref-literals.md:50]: shall we use `r` to be consistent with the example?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -1805,6 +1805,8 @@ HAT: '^';
 STRING
     : '\'' ( ~('\''|'\\') | ('\\' .) )* '\''
     | '"' ( ~('"'|'\\') | ('\\' .) )* '"'
+    | 'R\'' (~'\'')* '\''
+    | 'R"'(~'"')* '"'
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1809]: shall we define a new token for it? then the parser can choose to not call `unescapeSQLString` for raw string token.

File: docs/sql-ref-literals.md (hunk 1 above)
Comment (sarutak) [reply docs/sql-ref-literals.md]: Thanks. Updated.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (sarutak) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1813]: If we have a new token aside from `STRING`, we need to care all the places in `SqlBase.g4` and `ASTBuilder.scala` where `STRING` is referred. Is it acceptable?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala [hunk 3]
```diff
@@ -158,42 +158,46 @@ object ParserUtils {
...
-          sb.append(Integer.parseInt(cp, 8).toChar)
-          charBuffer.position(charBuffer.position() + 4)
-        case ESCAPED_CHAR_PATTERN(c) =>
-          // escaped character literals.
-          appendEscapedChar(c.charAt(0))
-          charBuffer.position(charBuffer.position() + 2)
-        case _ =>
-          // non-escaped character literals.
-          sb.append(charBuffer.get())
+    if (b.matches("""(r|R)('|").*""")) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:161]: is it more efficient to do `b.startsWith("r") || b.startsWith("R")`?

**Review by cloud-fan — APPROVED** [review]

**Comment by sarutak** [comment]
retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**